src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,233 @@
+        var savedFullReport = coverageReportRepository.save(fullReport);
+
+        // The file reports unique for a file path. the set is used to aggregate multiple file reports from multiple test
+        // cases into one file report
+        var uniqueFileReports = new HashSet<CoverageFileReport>();
+
+        fileReportByTestCaseName.forEach((testCaseName, fileReports) -> {
+            // retrieve the test matching the extracted test case name
+            var optionalTestCase = testCases.stream().filter(testCase -> testCaseName.equals(testCase.getTestName())).findFirst();
+            if (optionalTestCase.isEmpty()) {
+                return;
+            }
+            var testCase = optionalTestCase.get();
+
+            fileReports.forEach(fileReport -> {
+                // If the file does not exist in the solution repository, no file report will be created
+                // This is for example the case if the test itself invokes code in a test class
+                if (solutionLineCountByFilePath.get(fileReport.getFilePath()) == null) {
+                    return;
+                }
+
+                // Temporarily save the testwise entries for the current file report as the entries do not have
+                // an ID and therefore cause an exception when the file report is saved to the database
+                var testwiseEntries = fileReport.getTestwiseCoverageEntries();
+
+                var optionalReportWithSameName = uniqueFileReports.stream().filter(report -> report.getFilePath().equals(fileReport.getFilePath())).findFirst();
+                CoverageFileReport savedFileReport;
+
+                if (optionalReportWithSameName.isEmpty()) {
+                    // Remove the testwise entries temporarily as they do not have an ID yet
+                    fileReport.setTestwiseCoverageEntries(Collections.emptySet());
+                    fileReport.setFullReport(savedFullReport);
+                    // Save the line count
+                    var lineCount = solutionLineCountByFilePath.get(fileReport.getFilePath());
+                    fileReport.setLineCount(lineCount);
+
+                    savedFileReport = coverageFileReportRepository.save(fileReport);
+                }
+                else {
+                    var reportWithSameName = optionalReportWithSameName.get();
+                    savedFileReport = coverageFileReportRepository.save(reportWithSameName);
+                }
+                uniqueFileReports.add(savedFileReport);
+
+                // Save all entries for the current file report to the database
+                testwiseEntries.forEach(entry -> {
+                    entry.setTestCase(testCase);
+                    entry.setFileReport(savedFileReport);
+                    testwiseCoverageReportEntryRepository.save(entry);
+                });
+            });
+        });
+
+        var updatedFullReport = coverageReportRepository.findByIdElseThrow(savedFullReport.getId());
```

Comment:
Three reports are needed if streams are used (which I highly prefer over for-loops) because variables inside the streams need to be (effectively ) final.